197 update rke2 installation instructions (#235)
defenseunicorns/uds-bundle-software-factory-nutanix: README.md

# UDS Bundled Software Factory for Nutanix

> This is a UDS Software Facotry bundle meant to be deployed on a Nutanix-based K8s cluster. This is not a [Supported Environment](https://uds.defenseunicorns.com/reference/uds-core/distribution-support/) but UDS should be compatible with this setup, and this bundle includes configuration changes to help it work on Nutanix.

Reviewer: ```suggestion
This is a UDS Software Factory bundle intended to be deployed on an RKE2 K8s cluster running in Nutanix. This bundle utilizes the nutanix-csi driver to support running UDS Software Factory on Nutanix. It has been tested using RKE2 clusters deployed with the [Nutanix RKE2 module](https://github.com/defenseunicorns/delivery-nutanix-iac/tree/main/modules/rke2).
```

Reviewer: Updated language to clarify the distro/infra that this bundle is designed to be used on. Also replaced language about Nutanix not being a supported environment with a more clear statement about the configuration in place that makes the bundle support Nutanix followed by a statement calling out the IaC used to generate the supported environments.

Author: Took the suggestion largely, dropped the comment about the CSI driver as unnecessary, it's in the bundle, they don't need to think about it.

# UDS Bundled Software Factory for Nutanix

This is a UDS Software Factory bundle intended to be deployed on an RKE2 K8s cluster running in Nutanix. This bundle utilizes the nutanix-csi driver to support running UDS Software Factory on Nutanix. It has been tested using RKE2 clusters deployed with the [Nutanix RKE2 module](https://github.com/defenseunicorns/delivery-nutanix-iac/tree/main/modules/rke2).

A list of included applications can be found [here](docs/packages-and-dependencies.md).

Bundle developers see [development.md](docs/development.md).

## Installing on Nutanix

> For further insight into the underlying infrastructure including the RKE2 cluster, see [the Nutanix IAC repo](https://github.com/defenseunicorns/delivery-nutanix-iac/tree/main). In particular the [UDS SWF pre-requisites documentation](https://github.com/defenseunicorns/delivery-nutanix-iac/blob/main/docs/uds-swf-prereqs.md). Note that as this bundle continues to be improved the above document may not perfectly capture the required buckets and databases. Also, buckets cannot be created in Nutanix via Terraform/Tofu. Those are always done via click-ops.

Reviewer: ```suggestion
For further insight into the underlying infrastructure including the RKE2 cluster, see [the Nutanix IAC repo](https://github.com/defenseunicorns/delivery-nutanix-iac/tree/main). In particular the [UDS SWF pre-requisites documentation](https://github.com/defenseunicorns/delivery-nutanix-iac/blob/main/docs/uds-swf-prereqs.md). Note that as this bundle continues to be improved the above document may not perfectly capture the required buckets and databases. Also, buckets cannot be created in Nutanix Objects via Terraform/Tofu. Those must be created via click-ops.
```

### 1. Install Required Tools

- [UDS CLI](https://github.com/defenseunicorns/uds-cli/tree/v0.16.0). This links to version 0.16.0 which is what the bundle is tested with but it should work with most nearby versions. The binary is here: <https://github.com/defenseunicorns/uds-cli/releases/download/v0.16.0/uds-cli_v0.16.0_Linux_amd64>

> **Optionally**, the following tools are baked into the UDS CLI via Zarf (`uds zarf tools`) already are easier to use if installed directly:

Reviewer: I have no problem still referencing these, but I think they should be mentioned as recommended tools for interacting/troubleshooting a bundle that has been deployed rather than calling them "optional" tools for deploying a bundle. The only cli tool truly required is uds-cli. Will let you determine a better way to reference them.

Author: Ah, good catch.

- [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl)
- [helm](https://helm.sh/docs/intro/install/)
- [k9s](https://k9scli.io/topics/install/) also available via the command `uds monitor`

### 2. Get the Bundle

Get the bundle you would like to install:
  - You can download the latest in your browser from [Defense Unicorn's published packages](https://github.com/orgs/defenseunicorns/packages?repo_name=uds-bundle-software-factory-nutanix)
  - You can pull it via the same protocol used to push/pull docker images:
    ```bash

>     uds pull oci://ghcr.io/defenseunicorns/uds-bundle/software-factory-nutanix-rke2:0.4.3 --architecture amd64`

Reviewer: Is there someway to have release please update these references in the readme when it releases a new version? They already don't match the latest version and it would be nice for them to autoupdate to the latest version

Reviewer: If that proves too complicated to update in the README, don't let it block the PR. But if it is simple it would be nice to add

Author: Yes it might be possible. I'm not totally sure.

    ```

>   - You can also reference it by it's "docker name" (OCI image URL) at deploy time with the same command above, `pull` switched to `deploy` and UDS will pull it and deploy it. This is less relevant to air-gapped installs unless you're pushing the bundle to a high-side docker registry before install (which isn't a bad idea).

Reviewer: ```suggestion
  - You can also reference it by it's "docker name" (OCI image URL) at deploy time with the same command above, `pull` switched to `deploy` and UDS will pull it and deploy it. This example is using the OCI published to the github container registry. For an air-gapped install this could be done by pushing the bundle into a high-side OCI registry and referencing the image by the hostname or IP of that registry.
```

    ```bash
    uds deploy oci://ghcr.io/defenseunicorns/uds-bundle/software-factory-nutanix-rke2:0.4.3 --architecture amd64`
    ```

### 3. Customize Configuration via `uds-config.yaml`

Create your own `uds-config.yaml` file. You can start from the reference file in the `config/` directory. Note it the default yaml file includes comments to point you at particular docs (such as [docs/nexus.md](docs/nexus.md)) to guide you in understanding and configuring some of the more challenging values.

Don't worry about the domain, database, TLS, and object store related values at this time. They will be addressed in the following sections.

### 4. Create Infrastructure Dependencies

This bundle requires pre-existing s3 buckets and external postgres databases. The addresses and credentials are passed in via the `uds-config.yaml` file at deploy time.

#### A. Create the Kubernetes Cluster

> Goes without saying, but the k8s cluster must already exist in Nutanix, and you must have admin access to it. We create ours with Terraform [here](https://github.com/defenseunicorns/delivery-nutanix-iac/tree/main)

Author: pushed a modified version

#### B. Provide a Domain, DNS sub-domain support (wildcard CNAME preferable), TLS certs

> During bundle development and testing, we use the `*.bigbang.dev` domain (which is owned by a Unicorn who used to work for Platform One). You will need to setup routing to `*.your-domain.com`.

Reviewer: ```suggestion
During bundle development and testing, we use the `*.bigbang.dev` domain. You will need to setup DNS entries for `*.your-domain.com` and `*.nexus.your-domain.com` to resolve the tenant load balancer IP and `*.admin.your-domain.com` to resolve the admin load balancer IP. These IPs depend on the IP range provided to the `metallb.IP_ADDRESS_POOL` variable in the uds-config.yaml file. Metallb assigns IPs in order as requested, and the admin gateway is created first, so the admin gateway will be the first IP in the range and the tenant gateway IP will be the second IP in the range provided. Using the values in the example uds-config.yaml, the admin gateway would be assigned IP 10.0.0.10 and the tenant gateway would be assigned 10.0.0.11.
```

Reviewer: Removed blurb about the owner of the bigbang.dev domain and expanded "setup routing" with more specific instructions. Feel free to reorganize or rephrase things, but trying to make it more clear what DNS entries need configured and how to find what to set them to.

  - `grafana.your.domain`
  - `neuvector.your.domain`
  - `nexus.your.domain`
  - `*.nexus.your.domain`

>     > **Note:** if you create a cert per nexus subdomain instead of a wildcard cert (subdomains are docker registries) see [docs/nexus.md](docs/nexus.md) to make sure you update the tenant gateway and other relevant uds-config variables correctly. It was tested with a wildcard cert. This author can't promise the current configuration would allow multiple individual certs to be inserted into the right places.

Reviewer: ```suggestion
    > **Note:** if you create your cert with a SAN per nexus subdomain instead of a wildcard SAN (need one subdomain per image registry) see [docs/nexus.md](docs/nexus.md) to make sure you update the tenant gateway and other relevant uds-config variables correctly. It was tested with a wildcard SAN on the tenant cert which enables creating multiple image registries without needing to perform a uds deploy to reconfigure networking. 
```

Author: OK, I'll do a search. I'm a bit concerned that the term "OCI" could confuse people who aren't well enough initiated to know what that is. Docker gets at the idea quickly like talking about a Kleenex brings to mind a tissue. I'll do a search for docker and at least get them to image/container image.

Done

Author: Pushed a modified version of this.

  - `sonarqube.your.domain`
  - `tracing.your.domain`

Update the domain and TLS cert values in your `uds-config.yaml` file.

> #### C. Provision S3-like Object Storage - TODO: ref via IAC

Reviewer: ```suggestion
#### C. Provision S3-like Object Storage
```
Removed TODO and reference to IaC that doesn't exist

There are the default bucket names in the default `uds-config.yaml` file. If you choose to deviate from these names know:

> - Gitlab only allows you to add a suffix.

Reviewer: ```suggestion
- The Gitlab package only allows you to provide a suffix that is added to the default names
```

Author: Disagree with specifying the GitLab package unless we know the helm chart would permit it otherwise. No need to make ourselves sound inflexible. Otherwise mostly agree. I think using the word suffix sort of implies the addition to the default names, so I dropped the word count there.

> - Loki, Velero and Mattermost don't care what you name the S3 buckets.

Reviewer: ```suggestion
- Loki, Velero and Mattermost buckets can be named anything
```

Reference the `uds-config.yaml` file you created as you go to be sure you're creating/have created buckets by the intended names. For a list of UDS config variables to be sure you update, see [docs/object-store-creation-and-configuration.md](docs/object-store-creation-and-configuration.md).

#### D. Create the Postgres Databases.

The following applications require an external database:
- Keycloak
- Gitlab
- Sonarqube
- Jira
- Confluence
- Mattermost
- Nexus

Update your `uds-config.yaml` file with the correct credentials and connection URLs. For a list of UDS config variables to be sure you update, see [docs/database-creation-and-configuration.md](docs/database-creation-and-configuration.md).

> You can review the terraform IaC we use for databases in our [delivery-nutanix-deployments](https://github.com/defenseunicorns/delivery-nutanix-deployments) repo.

Reviewer: Specifying that we are using NDB for postgres databases and linking the module instead of the deployment repo

#### E. Enable Nutanix CSI

This bundle utilizes the Nutanix CSI Helm chart for persistent storage. It needs Nutanix resources set up ahead of time.

1. Prism Element user and password for the CSI provider to connect to Prism Element. Update your `uds-config.yaml` file with the user credentials and the Prism Element IP/Hostname.

> 2. Nutanix Storage Container for RWO persistent volumes. Add the storage container name to the `uds-config.yaml`.
> 3. Nutanix File Server for RWX persistent volumes. **Be sure to configure the DNS records that it asks for.** Put the File Server name as it appears in Prism Element into your `uds-config.yaml`.

Reviewer: ```suggestion
2. Prism Central user and password for the CSI provider to connect to Prism Central. Update your `uds-config.yaml` file with the user credentials and the Prism Central IP/Hostname.
3. Nutanix Storage Container for RWO persistent volumes. Add the storage container name to the `uds-config.yaml`.
4. Nutanix File Server for RWX persistent volumes. **Be sure to configure the DNS records that it asks for.** Put the File Server name as it appears in Prism Element into your `uds-config.yaml`.
```

Reviewer: Added additional required configuration for nutanix-csi that was added with version 3.0.0
